src/Interpreters/Session.cpp

```diff
@@ -441,15 +454,21 @@ ContextMutablePtr Session::makeSessionContext(const String & session_name_, std:
     prepared_client_info.reset();
 
     /// Set user information for the new context: current profiles, roles, access rights.
-    if (user_id && !new_session_context->getAccess()->tryGetUser())
-        new_session_context->setUser(*user_id);
+    ///  In the TCP handler, the quota key is received in "addendum packet" after context creation --> skip initialization.
+    if (!new_session_context->getAccess()->tryGetUser())
+        new_session_context->setUser(*user_id, res_client_info.interface != ClientInfo::Interface::TCP);
 
     /// Session context is ready.
     session_context = std::move(new_session_context);
     named_session = new_named_session;
     named_session_created = new_named_session_created;
     user = session_context->getUser();
 
+    session_tracker_handle = session_context->getSessionTracker().trackSession(
+        *user_id,
+        { session_name_ },
+        session_context->getSettingsRef().max_sessions_for_user);
```

Comment:
Hm, seems like it will count one names http session multiple times. Please add a test with named http sessions (see `session_id` parameter)

Reply:
I will add a test with different sesssion IDs, but this should not be an issue because session id must be unique for all sessions on server:
 
```
ClickHouse HTTP server returned 500 Internal Server Error: Code: 373. DB::Exception: Session 1 is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.7.1.1)
```

Reply:
Done (Test added)

Reply:
The tricky part is that named http sessions can be reused. Of course, it's not allowed to use the same session concurrently, but it's totally fine to use it sequentially from different http requests. A new session will not be created for each request, the old one will be reused instead.

Reply:
I've just tested it locally, and found that it counts an http session only if there's a running query in this session. So it's possible to open an unlimited number of named http sessions (`max_sessions_for_user = 2`):
```
    ~/ch/test  echo "select 1" | curl -X POST "http://localhost:8123/?user=maxsessions&session_id=1" -d @-                                                                                                                        ✔ 
1
    ~/ch/test  echo "select 1" | curl -X POST "http://localhost:8123/?user=maxsessions&session_id=2" -d @-                                                                                                                        ✔ 
1
    ~/ch/test  echo "select 1" | curl -X POST "http://localhost:8123/?user=maxsessions&session_id=3" -d @-                                                                                                                        ✔ 
1
    ~/ch/test  echo "select 1" | curl -X POST "http://localhost:8123/?user=maxsessions&session_id=4" -d @-                                                                                                                        ✔ 
1
    ~/ch/test  echo "select 1" | curl -X POST "http://localhost:8123/?user=maxsessions&session_id=5" -d @-                                                                                                                        ✔ 
1
```
(now there are 5 sessions, they will be closed after `default_session_timeout` seconds)

But it will not be possible to use more than 2 sessions concurrently anyway, so maybe it's fine

Reply:
Hm, no, it's possible if you update the setting with a SET query: https://pastila.nl/?0120fcc0/a41fe96d1a48546e4a4939fec4fc1c7a

Reply:
Confirm [this](https://github.com/ClickHouse/ClickHouse/pull/51724#discussion_r1269556881) issue. I'm going to fix it in the next update of this PR.

A simple uncoditional call to `new_session_context->setUser(*user_id)` in `makeSessionContext` for named sessions fixes the issue (in my old code not upmerged with master), But I'm not sure if there are any pitfalls I haven't considered.

I got several merge conflicts with master and can't update the code right now. I hope to finish the update on Monday if I do not meet other issues.

Reply:
> A simple uncoditional call to new_session_context->setUser(*user_id) in makeSessionContext for named sessions fixes the issue

See https://github.com/ClickHouse/ClickHouse/pull/51724#discussion_r1270606629

Reply:
This issue is no longer actual, because `SET for max_sessions_for_user` is prohibited now.

A similar case with `ALTER SETTINGS PROFILE` works fine. See the test [tests/queries/0_stateless/02832_alter_max_sessions_for_user.sh](https://github.com/ClickHouse/ClickHouse/blob/9e817803a6282c0101163d33c9b7b79f8769b31d/tests/queries/0_stateless/02832_alter_max_sessions_for_user.sh)